src/Resource.js

```diff
@@ -310,29 +687,71 @@ export default class Resource {
    *     return model.find({ filter: { name: 'Lassy' } }),
    *   },
    * }).then(results => {
-   *   return results.user.push('pets', results.pets);
+   *   const { user, pets } = results;
+   *   return user.push('pets', pets);
    * }).then(user => {
-   *   const newPets = user.relationship('pets'); // will include pets with ids '1' and '2'
+   *   const newPets = user.relationship('pets'); // will include pets with name of 'Lassy'
    * });
    * ```
    *
    * @async
    * @param {String} relationship
-   * @param {(String[]|ResourceArray)} ids
+   * @param {(String|String[]|Resource|ResourceArray)} data
    * @return {Promise<Resource>}
    */
   push(relationship, data) {
+    const relationshipObject = this.relationship(relationship);
+    const { relation, inverse, field, records } = relationshipObject;
+    const { schema, id, conn } = this;
+
+    let idsToPush;
+
+    if (relation !== 'hasMany') {
+      throw new TypeError(
+        `Tried calling 'push' on a resource whose 'relationship' is '${relation}'. This method ` +
+        'only works for resources whose \'relationship\' is \'hasMany\'.'
+      );
+    }
+
+    if (!isDataValidForSpliceAndPush(data)) {
+      throw new TypeError(
+        'Tried calling \'push\' with \'data\' that was neither a Resource, ResourceArray, ' +
+        'Array of Strings, nor a String ID.'
+      );
+    }
+
+    const pushData = (isCompliant) => {
```

Comment:
This function name should be extended because this function actually does two things:

1. Ensures `isCompliant` is true
2. Performs the push

The name should reflect what it's actually doing. Maybe `pushDataIfCompliant`? That would read well in the `.then` chaining below as well.